src/python/pants/build_graph/address.py

```diff
@@ -133,11 +115,45 @@ def parse(cls, spec, relative_to=''):
     spec_path, target_name = parse_spec(spec, relative_to=relative_to)
     return cls(spec_path, target_name)
 
+  @classmethod
+  def sanity_check(cls, spec_path, target_name):
+    def check_path(path):
+      # A root or relative spec is OK
+      if path == '':
+        return
+
+      normpath = os.path.normpath(path)
+      components = normpath.split(os.sep)
+      if components[0] in ('.', '..') or normpath != path:
+        raise ValueError("Spec has un-normalized path "
+                         "part '{path}'".format(path=path))
+      if components[-1].startswith('BUILD'):
+        raise ValueError('Spec path {path} has {trailing} as the last path part and BUILD is '
+                         'reserved files'.format(path=path, trailing=components[-1]))
+      if os.path.isabs(path):
+        raise ValueError('Spec has absolute path {path}; expected a path relative '
+                         'to the build root.'.format(path=path))
+
+    def check_target_name(name):
+      if not name:
+        raise ValueError('Spec {spec}:{name} has no name part'.format(spec=spec_path, name=name))
+
+      banned_chars = list(set(BANNED_CHARS_IN_TARGET_NAME) & set(name))
```

Comment:
It isn't necessary to create two sets here: a linear walk through the name, looking up each member in the second likely be faster?

Either way, should make BANNED_CHARS_IN_TARGET_NAME a set, rather than re-creating the set each time here.